dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java

```diff
@@ -0,0 +1,236 @@
+        for (int x = 0; x < index; x += 2) {
+            if (contents[x] == format) {
+                index -= 2;
+                if (x != index) {
+                    contents[x] = contents[index];
+                    contents[x + 1] = contents[index + 1];
+                }
+                contents[index] = null;
+                contents[index + 1] = null;
+                return;
+            }
+        }
+    }
+
+    public Set<Class<?>> getContentFormats() {
+
+        Set<Class<?>> c = new HashSet<>();
+        for (int x = 0; x < index; x += 2) {
+            c.add((Class<?>)contents[x]);
+        }
+        return c;
+    }
+
+    public void setDestination(Destination d) {
+        put(Destination.class, d);
+    }
+
+    public void setExchange(Exchange e) {
+        this.exchange = e;
+    }
+
+    public void setId(String i) {
+        this.id = i;
+    }
+
+    public void setInterceptorChain(InterceptorChain ic) {
+        this.interceptorChain = ic;
+    }
+    public Object put(String key, Object value) {
+        if (contextCache != null) {
+            contextCache.put(key, value);
+        }
+        return super.put(key, value);
+    }
+    public Object getContextualProperty(String key) {
+        if (contextCache == null) {
+            calcContextCache();
+        }
+        return contextCache.get(key);
+    }
+    public Set<String> getContextualPropertyKeys() {
+        return contextCache.keySet();
+    }
+
+    private void calcContextCache() {
+        Map<String, Object> o = new HashMap<>();
+        Exchange ex = getExchange();
+        if (ex != null) {
+            Bus b = ex.getBus();
+            if (b != null && b.getProperties() != null) {
+                o.putAll(b.getProperties());
+            }
+            Service sv = ex.getService();
+            if (sv != null) {
+                o.putAll(sv);
+            }
+            Endpoint ep = ex.getEndpoint();
+            if (ep != null) {
+                EndpointInfo ei = ep.getEndpointInfo();
+                if (ei != null) {
+                    Map<String, Object> p = ep.getEndpointInfo().getBinding().getProperties();
+                    if (p != null) {
+                        o.putAll(p);
+                    }
+                    p = ep.getEndpointInfo().getProperties();
+                    if (p != null) {
+                        o.putAll(p);
+                    }
+                }
+                o.putAll(ep);
+            }
+            o.putAll(ex);
+        }
+        o.putAll(this);
+        contextCache = o;
+    }
+    public static void copyContent(Message m1, Message m2) {
+        for (Class<?> c : m1.getContentFormats()) {
+            m2.setContent(c, m1.getContent(c));
+        }
+    }
+
+    public void resetContextCache() {
+        if (contextCache != null) {
+            contextCache = null;
+        }
+    }
+
+    public void setContextualProperty(String key, Object v) {
```

Comment:
default access modifier replaced with public due to a compilation problem at WS security 3.4.1 bundle